File: beginner/pagina08/2632/2632.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

struct Spell {
    string nome;
    int dano;
    vector<int> nivel;
}

Spell spells[] = {
    {"fire",  200,  {20, 30, 50}},
    {"earth", 400,  {25, 55, 70}},
    {"water", 300,  {10, 25, 40}},
    {"air",   100,  {18, 38, 60}}
};

int main(){
    int t;
    cin >> t;

    for(int i = 0; i < t; i++){
        int w, h, x0, y0;
        cin >> w >> h >> x0 >> y0;

        string magia;
        int nivel, cx, cy;
        cin >> magia >> nivel >> cx >> cy;

        Spell spell = {"", 0, {}};
        for(const auto& s: spells){
            if(s.nome == magia){
                spell = s;
                break;
            }
        }

        int dano = spell.dano;
        int raio = spell.nivel[nivel - 1];

        int esquerda = x0, direita = x0 + w, baixo = y0, cima = y0 + h;

        int ponto_mais_proximo_x = max(esquerda, min(cx, direita));
        int ponto_mais_proximo_y = max(baixo, min(cy, cima));

        double distancia = 
        (ponto_mais_proximo_x - cx) * (ponto_mais_proximo_x - cx) +
        (ponto_mais_proximo_y - cy) * (ponto_mais_proximo_y - cy);

        if(distancia <= raio){
            cout << dano << endl;
        } else {
            cout << 0 << endl;
        }
    }

    return 0;
}

File: beginner/pagina08/2632/2632.py
import math

def verificar_interseccao_circulo_retangulo(x0, y0, largura, altura, cx, cy, raio):
    x1, y1 = x0 + largura, y0 + altura

    ponto_mais_proximo_X = max(x0, min(cx, x1))
    ponto_mais_proximo_Y = max(y0, min(cy, y1))

    distancia = math.sqrt(
        (ponto_mais_proximo_X ** 2) + (ponto_mais_proximo_Y ** 2)
    )

    return distancia <= raio

livro_de_magias = {
    "fire":  ["dano": 200, "nivel": [20, 30, 50]],
    "earth": ["dano": 400, "nivel": [25, 55, 70]],
    "water": ["dano": 300, "nivel": [10, 25, 40]],
    "air":   ["dano": 100, "nivel": [18, 38, 60]],
}

casos_teste = int(input())

for _ in range(casos_teste):
    largura, altura, x0, y0 = map(int, input().split())
    magia, nivel, cx, xy = input().split()

    nivel = int(nivel) - 1, cx = int(cx), cy = int(cy)

    raio = livro_de_magias[magia]["nivel"][nivel]
    dano = livro_de_magias[magia]["dano"]

    interseccao = verificar_interseccao_circulo_retangulo(x0, y0, largura, altura, cx, cy, raio)

    print(dano if interseccao else 0)
